src/protocols/der/decode.c

```diff
@@ -0,0 +1,2852 @@
+	[FR_DER_TAG_SEQUENCE]	      = { .constructed = FR_DER_TAG_CONSTRUCTED, .decode = fr_der_decode_sequence },
+	[FR_DER_TAG_SET]	      = { .constructed = FR_DER_TAG_CONSTRUCTED, .decode = fr_der_decode_set },
+	[FR_DER_TAG_PRINTABLE_STRING] = { .constructed = FR_DER_TAG_PRIMATIVE,
+					  .decode      = fr_der_decode_printable_string },
+	[FR_DER_TAG_T61_STRING]	      = { .constructed = FR_DER_TAG_PRIMATIVE, .decode = fr_der_decode_t61_string },
+	[FR_DER_TAG_IA5_STRING]	      = { .constructed = FR_DER_TAG_PRIMATIVE, .decode = fr_der_decode_ia5_string },
+	[FR_DER_TAG_UTC_TIME]	      = { .constructed = FR_DER_TAG_PRIMATIVE, .decode = fr_der_decode_utc_time },
+	[FR_DER_TAG_GENERALIZED_TIME] = { .constructed = FR_DER_TAG_PRIMATIVE,
+					  .decode      = fr_der_decode_generalized_time },
+	[FR_DER_TAG_VISIBLE_STRING]   = { .constructed = FR_DER_TAG_PRIMATIVE, .decode = fr_der_decode_visible_string },
+	[FR_DER_TAG_GENERAL_STRING]   = { .constructed = FR_DER_TAG_PRIMATIVE, .decode = fr_der_decode_general_string },
+	[FR_DER_TAG_UNIVERSAL_STRING] = { .constructed = FR_DER_TAG_PRIMATIVE,
+					  .decode      = fr_der_decode_universal_string },
+
+	[UINT8_MAX] = { .constructed = FR_DER_TAG_PRIMATIVE, .decode = NULL },
+};
+
+static int decode_test_ctx(void **out, TALLOC_CTX *ctx, UNUSED fr_dict_t const *dict)
+{
+	fr_der_decode_ctx_t *test_ctx;
+
+	test_ctx = talloc_zero(ctx, fr_der_decode_ctx_t);
+	if (!test_ctx) return -1;
+
+	test_ctx->tmp_ctx	  = talloc(test_ctx, uint8_t);
+	test_ctx->oid_value_pairs = false;
+
+	*out = test_ctx;
+
+	return 0;
+}
+
+static ssize_t fr_der_decode_boolean(TALLOC_CTX *ctx, fr_pair_list_t *out, fr_dict_attr_t const *parent, fr_dbuff_t *in,
+				     UNUSED fr_der_decode_ctx_t *decode_ctx)
+{
+	fr_pair_t *vp;
+	fr_dbuff_t our_in = FR_DBUFF(in);
+	uint8_t	   val;
+
+	size_t len = fr_dbuff_remaining(&our_in);
+
+	if (!fr_type_is_bool(parent->type)) {
+		fr_strerror_printf("Boolean found in non-boolean attribute %s of type %s", parent->name,
+				   fr_type_to_str(parent->type));
+		return -1;
+	}
+
+	/*
+	 * 	ISO/IEC 8825-1:2021
+	 * 	8.2 Encoding of a boolean value
+	 * 	8.2.1 The encoding of a boolean value shall be primitive.
+	 *       	The contents octets shall consist of a single octet.
+	 * 	8.2.2 If the boolean value is:
+	 *       	FALSE the octet shall be zero [0x00].
+	 *       	If the boolean value is TRUE the octet shall have any non-zero value, as a sender's option.
+	 *
+	 * 	11.1 Boolean values
+	 * 		If the encoding represents the boolean value TRUE, its single contents octet shall have all
+	 *		eight bits set to one [0xFF]. (Contrast with 8.2.2.)
+	 */
+	if (len != 1) {
+		fr_strerror_printf("Boolean has incorrect length (%zu). Must be 1.", len);
+		return -1;
+	}
+
+	if (unlikely(fr_dbuff_out(&val, &our_in) < 0)) {
+		fr_strerror_const("Insufficient data for boolean");
+		return -1;
+	}
+
+	if (unlikely(val != DER_BOOLEAN_FALSE && val != DER_BOOLEAN_TRUE)) {
```

Comment:
brackets around things.  It's not required by C, but it cam avoid typos

`if ((a == b) || (c == d))`